{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "82b46649",
   "metadata": {},
   "source": [
    "# Gaboid documentation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcb06231",
   "metadata": {},
   "source": [
    "The command line execution of the graboid software is divided into modes *Database*, *Mapping*, *Calibrate*, *Report*, *Classify*. Within the designated working directory, subdirectories *results*, *calibration*, *data*, *tmp* and *warnings*. A dictionary containing generated files is pickled as a*catalog.pickle* file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be08e332",
   "metadata": {},
   "source": [
    "## Database\n",
    "### Usage\n",
    "**graboid DATABASE \\-\\-work_dir \\-\\-taxon \\-\\-marker \\-\\-fasta \\-\\-bold \\-\\-ranks \\-\\-chunksize \\-\\-max_attempts \\-\\-mv \\-\\-keep_tmp**\n",
    "\n",
    "### Description\n",
    "In this step, the program builds the reference database to be used in the classification set. The reference database is made up of a fasta file and a taxonomic table containing the taxon of each sequence for a specified set of taxonomic tanks. The sequence file may be provided by the user as a pre generated fasta file (**\\-\\-fasta**) or built using a cross search of the genomic databases for a given taxon/marker pair (**\\-\\-taxon**, **\\-\\-marker**).\n",
    "\n",
    "### Arguments\n",
    "**\\-\\-work_dir** \\<path to working directory\\>\n",
    "\n",
    ">Working directory to contain the generated files. The value provided must remain constant for all steps that use the generated database.\n",
    "\n",
    "**-T**, **\\-\\-taxon** \\<taxon\\>\n",
    "\n",
    ">Taxon to be used in the cross search. Provided value should belong to a sufficiently high rank (phylum/class).\n",
    "\n",
    "**-M**, **\\-\\-marker** \\<marker\\>\n",
    "\n",
    ">Genetic marker or gene to be used in the cross search.\n",
    "\n",
    "**-F**, **\\-\\-fasta** \\<path to fasta file\\>\n",
    "\n",
    ">Fasta file containing the sequences to be included in the reference database. Overrides **--taxon** and **--marker** if a value is provided.\n",
    "\n",
    "**\\-\\-bold**\n",
    "\n",
    ">Toggle ussage of the BOLD database. If included, the cross search includes the BOLD database in adition to Genbank.\n",
    "\n",
    "**-r**, **\\-\\-ranks** \\<rank1\\> \\<rank2\\> ...\n",
    "\n",
    ">Ranks to be included in the retrieved taxonomic table. Default values are *phylum*, *class*, *order*, *family*, *genus* and *species*\n",
    "\n",
    "**-c**, **\\-\\-chunksize** = 500\n",
    "\n",
    ">The set of sequences to be downloaded is partitioned into segments of *chunksize* elements. Default value is 500.\n",
    "\n",
    "**-m**, **\\-\\-max_attempts** = 3\n",
    "\n",
    ">Determines the maximum number of failed attempts at retrieving a seuqnece chunk before writing it off as a failure.  The program attempts a second download run to retrieved failed sequences. Sequences that fail to download after this are listed in a warning file.\n",
    "\n",
    "**\\-\\-mv**\n",
    "\n",
    ">Toggles move of the provided fasta file. If enabled, the file provided as **--fasta** is moved to the working directory instead of copied.\n",
    "\n",
    "**\\-\\-keep_tmp**\n",
    "\n",
    ">Toggles deletion of temporal files. If provided, the generated temporal files are kept, otherwise, the temporal directory is wiped."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a0c89bb",
   "metadata": {},
   "source": [
    "## Mapper\n",
    "### Usage\n",
    "**graboid MAPPER \\-\\-work_dir \\-\\-base_seq \\-\\-db_dir \\-\\-out_name \\-\\-ref_name \\-\\-evalue \\-\\-threads**\n",
    "\n",
    "### Description\n",
    "Align the sequences retrieved in the *DATABASE* step against a baseline sequence of the specified marker and process result into a numeric matrix. Additionally, quantifies information present in each site and generates an order file containing the sites for each taxon sorted by descending information content.\n",
    "\n",
    "### Arguments\n",
    "**\\-\\-work_dir** \\<path to working directory\\>\n",
    "\n",
    ">Working directory to contain the generated files. The value provided must remain constant for all steps that use the generated database.\n",
    "\n",
    "**-B**, **\\-\\-base_seq** \\<base sequence file\\>\n",
    "\n",
    ">Marker sequence to be used as base of the sequence alignment.\n",
    "\n",
    "**-db**, **\\-\\-db_dir** \\<path to blast database directory\\>\n",
    "\n",
    ">OPTIONAL. If a blast database of the baseline sequence is already available, it may be provided in place of **--base_seq**\n",
    "\n",
    "**-o**, **\\-\\-out_name** \\<output file name\\>\n",
    "\n",
    ">OPTIONAL. Alternative name for the generated files.\n",
    "\n",
    "**-bn**, **\\-\\-blast_name** \\<blast report file name\\>\n",
    "\n",
    ">OPTIONAL. Alternative name for the blast report.\n",
    "\n",
    "**-e**, **\\-\\-evalue** = 0.005\n",
    "\n",
    ">E-value threshold for the blast alignment. Default value is 0.005.\n",
    "\n",
    "**-t**, **\\-\\-threads** = 1\n",
    "\n",
    ">Number of threads to use in the blast alignment. Default value is 1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6ea010d",
   "metadata": {},
   "source": [
    "## Calibrate\n",
    "### Usage\n",
    "**graboid CALIBRATE \\-\\-work_dir \\-\\-row_thresh \\-\\-col_thresh \\-\\-min_seqs \\-\\-rank \\-\\-dist_mat \\-\\-w_size \\-\\-w_step \\-\\-max_k \\-\\-step_k \\-\\-max_n \\-\\-step_n \\-\\-min_k \\-\\-min_n \\-\\-out_file**\n",
    "\n",
    "### Description\n",
    "Perform a grid search for combinations of *window*, *k* and *n* values using the *leave-one-out* method to assess performance for each taxon. Generates a calibration report containing the *Accuracy*, *Precision*, *Recall* and *F1 score* for each taxon in each parameter combination.\n",
    "\n",
    "### Arguments\n",
    "**\\-\\-work_dir** \\<path to working directory\\>\n",
    "\n",
    ">Working directory to contain the generated files. The value provided must remain constant for all steps that use the generated database.\n",
    "\n",
    "**-rt**, **\\-\\-row_thresh** = 0.2\n",
    "\n",
    ">Empty row threshold. Percentage threshold of empty values to filter an alignment row. Default value is 0.2.\n",
    "\n",
    "**-ct**, **\\-\\-col_thresh** = 0.2\n",
    "\n",
    ">Empty column threshold. Percentage threshold of empty values to filter an alignment column. Default value is 0.2.\n",
    "\n",
    "**-ms**, **\\-\\-min_seqs** = 10\n",
    "\n",
    ">Minimum number of sequences needed for a taxon to be considered. Default value is 10.\n",
    "\n",
    "**-rk**, **\\-\\-rank** = 'genus'\n",
    "\n",
    ">Taxonomic rank used for attribute selection. Default value is *genus*.\n",
    "\n",
    "**-dm**, **\\-\\-dist_mat** \\<distance matrix\\>\n",
    "\n",
    ">Distance matrix to be used in the neighbour calculation. String *id* to use the identity matrix. String *s1v2* to use the transition = 1, transversion = 2 matrix. A custom matrix can be used by passing it as a file path.\n",
    "\n",
    "**-wz**, **\\-\\-w_size** = 200\n",
    "\n",
    ">Window size to be used in each iteration. Default value is 200.\n",
    "\n",
    "**-ws**, **\\-\\-w_step** = 15\n",
    "\n",
    ">Window displacement between iterations. Default value is 15.\n",
    "\n",
    "**-mk**, **\\-\\-max_k**\n",
    "\n",
    ">Maximum *k* value to be explored. Default value is 15.\n",
    "\n",
    "**-sk**, **\\-\\-step_k**\n",
    "\n",
    ">Increment in *k* value between iterations. Default value is 2.\n",
    "\n",
    "**-mn**, **\\-\\-max_n**\n",
    "\n",
    ">Maximum *n* value to be explored. Default value is 30.\n",
    "\n",
    "**-sn**, **\\-\\-step_n**\n",
    "\n",
    ">Increment in *n* value between iterations. Default value is 5.\n",
    "\n",
    "**-nk**, **\\-\\-min_k**\n",
    "\n",
    ">Starting *k* value. Default value is 1.\n",
    "\n",
    "**-nn**, **\\-\\-min_n**\n",
    "\n",
    ">Starting *n* value. Default value is 5.\n",
    "\n",
    "**-o**, **\\-\\-out_file**\n",
    "\n",
    ">OPTIONAL. File name for the generated report. If none is provided, the file name will be generated using the calibration parameters."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ccb0d99",
   "metadata": {},
   "source": [
    "## Report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cf41278",
   "metadata": {},
   "source": [
    "## Classify\n",
    "### Usage\n",
    "**graboid CLASSIFY \\-\\-work_dir \\-\\-query_file \\-\\-dist_mat \\-\\-w_start \\-\\-w_end \\-\\-k \\-\\-n \\-\\-cl_mode \\-\\-rank \\-\\-out_name \\-\\-keep_tmp**\n",
    "\n",
    "### Description\n",
    "Classify the sequences in the file given as **\\-\\-query_fasta**.\n",
    "\n",
    "### Arguments\n",
    "**--work_dir** \\<path to working directory\\>\n",
    "\n",
    ">Working directory to contain the generated files. The value provided must remain constant for all steps that use the generated database.\n",
    "\n",
    "**-q**, **\\-\\-query_file** \\<path to query sequence file\\>\n",
    "\n",
    ">File containing the query sequences. Fasta format.\n",
    "\n",
    "**-dm**, **\\-\\-dist_mat** \\<distance matrix\\>\n",
    "\n",
    ">Distance matrix to be used in the neighbour calculation. String id to use the identity matrix. String s1v2 to use the transition = 1, transversion = 2 matrix. A custom matrix can be used by passing it as a file path.\n",
    "\n",
    "**-ws**, **\\-\\-w_start** \\<window start\\>\n",
    "\n",
    ">Starting coordinates for the window of the alignment to use in classification.\n",
    "\n",
    "**-we**, **\\-\\-w_end** \\<window end\\>\n",
    "\n",
    ">End coordinates for the window of the alignment to use in classification.\n",
    "\n",
    "**\\-\\-k** \\<k1\\> \\<k2\\>...\n",
    "\n",
    ">K values to use in classification. Multiple values can be provided to assess consistency in classification results.\n",
    "\n",
    "**\\-\\-n** \\<number of sites\\>\n",
    "\n",
    ">Number of informative sites to use in the classification\n",
    "\n",
    "**-md**, **\\-\\-cl_mode** \\<classification mode\\>\n",
    "\n",
    ">Classification criterion to be used. Single character string: *m* to use majority mode classification, *w* to use wKNN classification and *d* to use dwKNN classification.\n",
    "\n",
    "**-rk**, **\\-\\-rank** = 'genus'\n",
    "\n",
    ">Taxonomic rank used for attribute selection. Default value is *genus*.\n",
    "\n",
    "**-o**, **\\-\\-out_file**\n",
    "\n",
    ">OPTIONAL. File name for the generated results. If none is provided, the filename is generated from datetime and the query file name.\n",
    "\n",
    "**\\-\\-keep_tmp**\n",
    "\n",
    ">Toggles deletion of temporal files. If provided, the generated temporal files are kept, otherwise, the temporal directory is wiped."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
