{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "15f5febe",
   "metadata": {},
   "source": [
    "# Graboid documentation\n",
    "## Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4ba3962",
   "metadata": {},
   "source": [
    "### BLAST\n",
    "This module handles the alignment of the database or query sequences to a single reference sequence of the selected molecular marker."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b9d85ea",
   "metadata": {},
   "source": [
    "#### Functions\n",
    "* **blast(query, ref, out_dir, threads = 1)** Perform an ungapped blastn search of the given *query* against the given *ref* result is stored to *out_file*\n",
    "* **makeblastdb(ref_file, db_prefix)** Make a blast database for the given *ref_file*. Generated files use the given *db_prefix*\n",
    "* **check_db_dir(db_dir)** Counts possible blastdb files in the directory *db_dir*. Returns True if exactly 6 files are found, False otherwise. Additionally returns a list of the found files\n",
    "\n",
    "#### Blaster\n",
    "**class Blaster()**\n",
    "\n",
    "This class handles all blast related tasks\n",
    "\n",
    "##### Attributes\n",
    "* **report** Path to the generated blast report\n",
    "\n",
    "##### Methods\n",
    "* **make_ref_db(ref_file, db_dir, clear = False)** Generates a blast database for the given *ref_file*. Generated files are stored in the directory *db_dir*. If the specified directory already contains database files and the argument *clear* is set to true, delete existing files and replace them with new ones.\n",
    "* **blast(fasta_file, out_file, db_dir, threads = 1)** Blasts the sequences of *fasta_file* against the database given by *db_dir*. Attribute *report* updates with the path to the generated file (if BLAST is successful)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc9d6ca2",
   "metadata": {},
   "source": [
    "### Matrix2 (rename to matrix_builder)\n",
    "This module builds an alignment matrix from the blast report generated by a *blaster* instance."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff707bc0",
   "metadata": {},
   "source": [
    "#### Variables\n",
    "* **bases**\n",
    "* **tr_dict**\n",
    "\n",
    "#### Functions\n",
    "* **read_blast(blast_file, evalue = 0.005)** Loads the blast report stored in *blast file*. Assumes the columns generated are *qseqid*, *pident*, *length*, *qstart*, *qend*, *sstart*, *send* and *evalue*. Max e-value allowed is determined by the argument *evalue*. Returns loaded table as a pandas dataframe.\n",
    "* **read_seqfile(seq_file)** Loads a fasta file sepecified by *seq_file*. Returns a dictionary with accession:sequence as key:value pairs\n",
    "* **get_seqs(seq_file, blast_tab)** Returns a dictionary containing the sequences from *seq_file* present in *blast_tab*. Key:value pairs are accession:sequence\n",
    "* **get_mat_dims(blast_tab)** Infers the alignment matrix dimensions from the values in *blast_tab*. Returns *nrows*, taken as the number of rows in *blast_tab*, *ncols*, as the maximum extent of matches over the reference sequence, and *offset* as the minimum value of match starts in *blast_tab*\n",
    "* **build_coords0(coord_mat)** Sorts coordinates for the given match. Flips reverse matches. Returns *seq_coords*, the portions of each sequence to take, and *mat_coords* the location of each match in the alignment matrix.\n",
    "* **build_row0(seq, seq_coords, mat_coords, rowlen)** Uses *seq_coords* and *mat_coords* to incorporate *seq* into a row of the alignment matrix. Returns *row*, a numpy array of length *rowlen* containing the aligned *seq* turned to numeric values.\n",
    "* **plot_coverage_data(blast_file, evalue = 0.005, figzise = (12, 7))** Plots sequence coverage taken from the blast report. X axis shows alignment coordinates.\n",
    "\n",
    "#### MatBuilder\n",
    "**class MatBuilder(out_dir)**\n",
    "\n",
    "##### Parameters\n",
    "* **out_dir** Output directory for the generated alignment file\n",
    "\n",
    "##### Attributes\n",
    "* **out_dir**\n",
    "* **dims** Dimensions of the generated alignment matrix (*cols*, *rows*)\n",
    "* **acclist** List of accessions included in the alignment (used to locate a given sequence in the matrix)\n",
    "* **mat_file** File name of the generated alignment matrix. Suffix *.npy*\n",
    "* **acc_file** File name of the generated accession list. Suffix *.acclist*\n",
    "\n",
    "##### Methods\n",
    "* **generate_outnames(seq_file, out_name = None)** Builds the file names for the generated files. If *out_name* is provided, generated file names are *\\<out_name>.npy* and *\\<out_name>.acclist*\n",
    "* **build(blast_file, seq_file, out_name = None, evalue = 0.005)** Use the blast report given in *blast_file* to generate an alignment matrix from the data contained in *seq_file*. Only the matches below the *evalue* threshold are included in the alignment matrix. The method generates output file names on runtime (names can be specified with the argument *out_name*)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd0db214",
   "metadata": {},
   "source": [
    "### Director\n",
    "This module handles the blast search and alignment matrix construction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "754e420f",
   "metadata": {},
   "source": [
    "#### Functions\n",
    "* **make_dirs(base_dir)** Generates the necesary subdirectories in base_dir to contain the generated files. Subdirectory names are data and warnings.\n",
    "* **check_fasta(fasta_file)** Returns the count of recognized fasta sequences in *fasta_file*. Used to verify format of input sequence files.\n",
    "\n",
    "#### Director\n",
    "**class Director(out_dir, warn_dir)**\n",
    "\n",
    "This class handles the process of the blast search and matrix construction. Stores paths to generated files and matrix metadata.\n",
    "##### Parameters\n",
    "* **out_dir** Directory for the output sequence, accession, taxonomy and taxonomy guide files\n",
    "* **warn_dir** Directory for the warning files generated along the process\n",
    "\n",
    "##### Attributes\n",
    "* **out_dir**\n",
    "* **warn_dir**\n",
    "* **warn_handler**\n",
    "* **log_handler**\n",
    "* **db_dir** Directory containing the BLAST database files\n",
    "* **blast_report** Path to the generated BLAST report\n",
    "* **mat_file** Path to the generated alignment matrix\n",
    "* **acc_file** Path to the generated accession list\n",
    "* **dims** Dimensions of the generated alignment matrix\n",
    "* **workers** The workers are instances of the classes used to construct the sequence map\n",
    "  * **blaster**\n",
    "  * **mapper**\n",
    "\n",
    "##### Methods\n",
    "* **set_blastdb(db_dir)** Establish the given *db_dir* as the BLAST database directory. Note, specified directory must contain the databse files.\n",
    "* **build_blastdb(ref_seq, ref_name = None)** Creates a blast database using the given *ref_seq* (NOTE: ref seq must contain a single sequence). If no *ref_name* is given, the output directory is named based on the *ref_seq* file.\n",
    "* **direct(fasta_file, out_name = None, evalue = 0.005, threads = 1)** Perform the BLAST alignment and build the corresponding matrix. NOTE: a BLAST database must be set beforehand. Perform a blast alginment of the given *fasta_file*. If *out_name* is provided, the generated report is saved in *\\<out_name>.BLAST*, otherwise, the report is named after *fasta_file*. After the BLAST search, the report is used to generate the alignment matrix and accession list, these are stored to files *\\<out_name>.npy* and *\\<out_name>.acclist* respectively."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
